lib/ai/chat.go

```diff
@@ -0,0 +1,250 @@
+	Error <-chan error
+}
+
+// Complete completes the conversation with a message from the assistant based on the current context.
+// On success, it returns the message and the number of tokens used for the completion.
+func (chat *Chat) Complete(ctx context.Context) (any, int, error) {
+	var numTokens int
+
+	// if the chat is empty, return the initial response we predefine instead of querying GPT-4
+	if len(chat.messages) == 1 {
+		return &Message{
+			Role:    openai.ChatMessageRoleAssistant,
+			Content: initialAIResponse,
+			Idx:     len(chat.messages) - 1,
+		}, numTokens, nil
+	}
+
+	// if not, copy the current chat log to a new slice and append the suffix instruction
+	messages := make([]openai.ChatCompletionMessage, len(chat.messages)+1)
+	copy(messages, chat.messages)
+	messages[len(messages)-1] = openai.ChatCompletionMessage{
+		Role:    openai.ChatMessageRoleUser,
+		Content: promptExtractInstruction,
+	}
+
+	// create a streaming completion request, we do this to optimistically stream the response when
+	// we don't believe it's a payload
+	stream, err := chat.client.svc.CreateChatCompletionStream(
+		ctx,
+		openai.ChatCompletionRequest{
+			Model:     openai.GPT4,
+			Messages:  messages,
+			MaxTokens: maxResponseTokens,
+			Stream:    true,
+		},
+	)
+	if err != nil {
+		return nil, numTokens, trace.Wrap(err)
+	}
+
+	// fetch the first delta to check for a possible JSON payload
+	var response openai.ChatCompletionStreamResponse
+top:
+	response, err = stream.Recv()
+	if err != nil {
+		return nil, numTokens, trace.Wrap(err)
+	}
+	numTokens++
+
+	trimmed := strings.TrimSpace(response.Choices[0].Delta.Content)
+	if len(trimmed) == 0 {
+		goto top
+	}
+
+	// if it looks like a JSON payload, let's wait for the entire response and try to parse it
+	if strings.HasPrefix(trimmed, "{") {
+		payload := response.Choices[0].Delta.Content
+	outer:
+		for {
+			response, err := stream.Recv()
+			switch {
+			case errors.Is(err, io.EOF):
+				break outer
+			case err != nil:
+				return nil, numTokens, trace.Wrap(err)
+			}
+			numTokens++
+
+			payload += response.Choices[0].Delta.Content
+		}
+
+		// if we can parse it, return the parsed payload, otherwise return a non-streaming message
+		var c CompletionCommand
+		err = json.Unmarshal([]byte(payload), &c)
+		switch err {
+		case nil:
+			return &c, numTokens, nil
+		default:
+			return &Message{
+				Role:    openai.ChatMessageRoleAssistant,
+				Content: payload,
+				Idx:     len(chat.messages) - 1,
+			}, numTokens, nil
+		}
+	}
+
+	// if it doesn't look like a JSON payload, return a streaming message to the caller
+	chunks := make(chan string, 1)
+	errCh := make(chan error)
+	chunks <- response.Choices[0].Delta.Content
+	go func() {
+		defer close(chunks)
+
+		for {
+			response, err := stream.Recv()
+			switch {
+			case errors.Is(err, io.EOF):
+				return
+			case err != nil:
+				errCh <- trace.Wrap(err)
```

Comment:
this is going to block until the caller receives from the error channel and not get cancelled if the context is cancelled.

Reply:
@xacrimon Is this by design?

Reply:
I noted this during dev, thinking it was fine since the worker goroutine will cancel and return an error on cancellation itself. Now, in hindsight this isn't going to catch some oddball case of that goroutine panicking and not exiting via an error path so we should fix it by selecting on the cancellator here too.